README.md

```diff
@@ -60,7 +60,19 @@ listening for API requests and provisioning volumes accordingly.
 
 ### Architecture and Operation
 
-The PMEM-CSI driver can operate in two different DeviceModes: LVM and Direct
+The PMEM-CSI driver can operate in two different DeviceModes: LVM and Direct. This tables describes shows overview and comparison of those modes. There is more detailed modes explanation down in following paragraphs.
+
+|                   |LVM                    |Direct              |
+|:--                |:--                    |:--                 |
+|Purpose            |avoids fragmentation   |simpler, somewhat faster, but storage may get fragmented   |
+|What is served     |LVM Logical Volume     |pmem block device   |
+|Region-affinity    |yes:One Volume Group per Region  |yes: namespace can belong to one Region only  |
+|Startup            |Two extra stages: pmem-ns-init(creates namespaces), vgm(creates Volume Groups)   |No extra steps at startup |
+|Namespace modes    |fsdax, sector -mode namespaces precreated as pools   |Namespace of required mode created directly, no need to pre-create pools   |
+|Limiting space use | Can leave part of device unused during pools creation  | No limits, create namespaces on device until runs out of space  |
+| *Name* field in Namespace | *Name* gets set to 'pmem-csi' to achieve own vs. foreign marking | *Name* gets set to VolumeID, without attempting own vs. foreign marking  |
+|Minimum volume size| 4MB                   | 2GB (See also alignment adjusting below) |
```

Comment:
Why 2GB? From the description below it sounds like 1GB is the minimum.

Reply:
during test and trials, I developed knowledge that I should not attempt allocation less than 2 GB. Thats because I learned at some point that libndctl allocates me 1-step less than I asked, and libndctl aligns down. But that was during the times when pmem-csi itself also aligned down, so worth rechecking now.

Reply:
Okay, so it is a minimum size that has nothing to do with alignment. The comment in brackets made me belief that.

We should indeed investigate whether it's still needed (would help our CI!) and/or rephrase the text: "at least 2GB, potentially more depending on alignment (see next row)".

Reply:
good that this triggered re-checking. It appears recent alignment related change actually broke direct-mode size calc, we now may allocate **less** than request (for example we now create 2 GB if we request 2.X GB).
Thats because in recent change I mixed alignment and size compensation together and over-simplified, taking away important "size += 2" line which was doing "combined align and compensate".
Compensate is needed because libndctl serves us 1GB less than we ask.
The more correct way, separating align and compensate layers, is to align up by 1GB and then add 1GB to compensate. (I will create separate PR).
But this compensation thing is entirely driver & libnsdctl-internal, so in user-facing terms our minimal is actually 1GB (what we should show in this table here), and that will be the served minimal actual size if we ask typical small test volume, like 1MB.

Reply:
Can factor out the math into a helper function and add unit tests? It looks like this code is tricky enough that it can't be done correctly without better test coverage.

"to compensate" - I don't understand what that means. Compensate for what?

> in user-facing terms our minimal is actually 1GB (what we should show in this table here)

The table should show the smallest volume size that can be created in this mode.

Reply:
> Compensate for what?

libndctl creates namespace which is one alignment step smaller than we ask. Thats I learned on path, but I should write it in big letters somewhere. It's easy to demonstrate on cmdline:
```
# ndctl create-namespace -s 2G -a 1G
{
  "dev":"namespace0.0",
  "mode":"fsdax",
  "map":"dev",
  "size":"1024.00 MiB (1073.74 MB)",
  "uuid":"35f3dad0-4650-48f5-93c4-0ebb2560e6b9",
  "sector_size":512,
  "align":1073741824,
  "blockdev":"pmem0"
}
```

and accordingly, attempt to create 1G will fail:
```
# ndctl create-namespace -s 1G -a 1G
libndctl: ndctl_pfn_enable: pfn0.0: failed to enable
  Error: namespace0.1: failed to enable
failed to create namespace: No such device or address
```

Reply:
> Can you factor out the math into a helper function and add unit tests?

good idea. yes, the pure math part can be extracted as separate function. However, this will not cover all interesting parts, like alignment forcing in CreateNamespace and size & alignment extra handling by libndctl, so this test will miss some quite important parts of the path. But sure, it is good idea to add such test and cover at least some part. Also, such test is convenient place to document and explain the rules behind the math through various sizes as test source data.